### PR TITLE
Implementation refactor for Client-based interface

### PR DESCRIPTION
**Pull request by fantix**
Review note: there is one merge commit "Partially revert ..." in this branch to reserve git history; use [this link](https://github.com/edgedb/edgedb-python/pull/286/files/5d4a0a5bd2e3ec16687fe075b189c44942983e14..1f51604e646e07eed343526ed8acaa17593362ae) to see the merged diff.

- [x] Merge Connection with RawConnection
- [x] Fix tests to use client
- [x] Add previously-deleted tests
- [ ] Complete timeout implementation and add more timeout tests (will be done in a new PR)
- [x] Drop `edgedb.connect()` and replace with a single-connection `Client`
- [x] Extract base class for Client, and further remove redundant code
- [x] Extract redundant code from transactions and only use `_iter_coroutine()` on top level
- [x] Review how pool is implemented, simplify if possible

edgedb-python-pool drawio

### CODE REVIEW

File: edgedb/blocking_client.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+        self._iter_coroutine(self._impl.expire_connections())
+
+    def __enter__(self):
+        return self.ensure_connected()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+
+def create_client(dsn=None, *, concurrency=None, **kwargs):
```
Comment (elprans) [inline edgedb/blocking_client.py:368]: Let's specify kwargs explicitly here for the sake of code self-documentation.

**Review by elprans — COMMENTED** [review]
Impressive work, Fantix!  A net reduction of ~700 lines in the implementation despite the addition of a thread-safe concurrent blocking pool.  I especially like the cleanup in the transaction implementation.

File: edgedb/blocking_client.py [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+        """Attempt to gracefully close all connections in the client.
+
+        Wait until all pool connections are released, close them and
+        shut down the pool.  If any error (including cancellation) occurs
+        in ``close()`` the pool will terminate by calling
+        Client.terminate() .
+        """
+        self._iter_coroutine(self._impl.close(timeout))
+
+    def expire_connections(self):
```
[the remark this answers is not in the document]
Comment (fantix) [reply edgedb/blocking_client.py]: No I don't think we need it.

File: tests/test_async_retry.py [hunk 3]
```diff
@@ -243,7 +243,8 @@ async def test_async_transaction_interface_errors(self):
                     async with tx:
                         pass
 
-        with self.assertRaisesRegex(edgedb.InterfaceError, r".*is borrowed.*"):
-            async for tx in self.con.transaction():
-                async with tx:
-                    await self.con.execute("SELECT 123")
+        # with self.assertRaisesRegex(edgedb.InterfaceError,
```
Comment (1st1) [inline tests/test_async_retry.py]: nit: let's make sure we remove this code in later commits -- we shouldn't have commented out code in the master branch.

**Review by 1st1 — APPROVED** [review]
Epic work, Fantix. I think this is ready to land.

**Pull request merged**